ref/java/src/main/java/org/bech32/Bech32.java

```diff
@@ -0,0 +1,154 @@
+package org.bech32;
+
+import org.apache.commons.lang3.tuple.Pair;
+
+public class Bech32 {
+
+    private static final String CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
+
+    public static String bech32Encode(byte[] hrp, byte[] data) {
+
+        byte[] chk = createChecksum(hrp, data);
+        byte[] combined = new byte[chk.length + data.length];
+
+        System.arraycopy(data, 0, combined, 0, data.length);
+        System.arraycopy(chk, 0, combined, data.length, chk.length);
+
+        byte[] xlat = new byte[combined.length];
+        for(int i = 0; i < combined.length; i++)   {
+            xlat[i] = (byte)CHARSET.charAt(combined[i]);
+        }
+
+        byte[] ret = new byte[hrp.length + xlat.length + 1];
+        System.arraycopy(hrp, 0, ret, 0, hrp.length);
+        System.arraycopy(new byte[] { 0x31 }, 0, ret, hrp.length, 1);
+        System.arraycopy(xlat, 0, ret, hrp.length + 1, xlat.length);
+
+        return new String(ret);
+    }
+
+    public static Pair<byte[], byte[]> bech32Decode(String bech) throws Exception  {
+
+        if(!bech.equals(bech.toLowerCase()) && !bech.equals(bech.toUpperCase()))  {
```

Comment:
Is there a risk that toLowerCase/toUpperCase are locale dependent?

Reply:
I don't think so given the character set being used. In any case, placing the test after the byte value test should remove any doubt.

Reply:
I'd still be a bit more confortable with `bech.toLowerCase(Locale.ROOT)`, which should be deterministic?

Reply:
Updated.